File: 03-herencia/java/Vehiculo.java
/*
 * Vehiculo.java - LA CLASE PADRE (superclase)
 *
 * Con los archivos separados se ve mejor por que existe "protected":
 * es un permiso que cruza de ESTE archivo a los archivos de los hijos.
 *
 *   private   -> solo Vehiculo.java
 *   protected -> Vehiculo.java + Auto.java + Moto.java (los hijos)
 *   public    -> todo el mundo
 */
public class Vehiculo {

    // protected = privado para el mundo, visible para los hijos
    protected String modelo;
    protected int anio;
    protected int velocidad;

    public Vehiculo(String modelo, int anio) {
        this.modelo = modelo;
        this.anio = anio;
        this.velocidad = 0;
    }

    public void acelerar() {
        this.velocidad += 20;
        System.out.println(this.modelo + " acelera a " + this.velocidad + " km/h");
    }

    public void describir() {
        System.out.println("Vehiculo generico: " + this.modelo + " (" + this.anio + ")");
    }
}

File: 03-herencia/python/vehiculo.py
"""
vehiculo.py - LA CLASE PADRE

En Python no existe "protected" de verdad. El guion bajo (self._modelo)
es solo una senal para el programador: "esto es interno, no lo uses
desde afuera". Los hijos lo usan igual, y el mundo exterior tambien
podria, aunque no deberia.
"""


class Vehiculo:

    def __init__(self, modelo, anio):
        self._modelo = modelo
        self._anio = anio
        self._velocidad = 0

    def acelerar(self):
        self._velocidad += 20
        print(f"{self._modelo} acelera a {self._velocidad} km/h")

    def describir(self):
        print(f"Vehiculo generico: {self._modelo} ({self._anio})")
